### ISSUE TITLE
Improve the GPU canary

### ISSUE THREAD

**Issue by goodfeli** [tensorflow/cleverhans#795]
I have a script in my personal repo called "upper_bound.py".
It's an upper bound on how much it takes to get a canary failure when the GPUs are having whatever sort of problem it is we keep encountering.
This script uses a simple weight decay loss, so I think we could make a model-free canary that just tries to do one step of synchronous gradient descent and sees if all the replicas agree. We'd want to make sure that we don't make a Variable that ends up in any collections. I think we could do this by just calculating synchronous gradients of an expression based on a constant weight tensor rather than weight decay on a Variable.
If we get such a canary working, we can use it to replace the current one. This would be nice because:
- The current canary has to be disabled whenever the loss function is stochastic, so we have no canary guarding jobs that involve training with dropout, training on noisy inputs, etc.
- We could put the new canary in non-training contexts. For example, we could put it the multi-GPU batch eval driver. This way we could have more confidence that our robustness benchmarking results aren't compromised by GPUs returning garbage results.
